src/lib/interactionManager/interaction.ts

```diff
@@ -0,0 +1,244 @@
+import { CredentialOfferFlow } from './credentialOfferFlow'
+import {
+  IdentitySummary,
+  CredentialVerificationSummary,
+} from '../../actions/sso/types'
+import { InteractionType } from 'jolocom-lib/js/interactionTokens/types'
+import {
+  JSONWebToken,
+  JWTEncodable,
+} from 'jolocom-lib/js/interactionTokens/JSONWebToken'
+import { BackendMiddleware } from '../../backendMiddleware'
+import {
+  InteractionChannel,
+  InteractionSummary,
+  SignedCredentialWithMetadata,
+} from './types'
+import { CredentialRequestFlow } from './credentialRequestFlow'
+import { JolocomLib } from 'jolocom-lib'
+import { CredentialMetadataSummary } from '../storage/storage'
+import { generateIdentitySummary } from 'src/actions/sso/utils'
+import { Flow } from './flow'
+import { last } from 'ramda'
+import { CredentialOfferRequest } from 'jolocom-lib/js/interactionTokens/credentialOfferRequest'
+import { AuthenticationFlow } from './authenticationFlow'
+import { CredentialRequest } from 'jolocom-lib/js/interactionTokens/credentialRequest'
+import { CredentialsReceive } from 'jolocom-lib/js/interactionTokens/credentialsReceive'
+import { Linking } from 'react-native'
+import { AppError, ErrorCode } from '../errors'
+
+/***
+ * - initiated by InteractionManager when an interaction starts
+ * - handles the communication channel of the interaction
+ * - holds the instance of the particular interaction (e.g. CredentialOffer, Authentication)
+ *
+ * TODO needs to hold:
+ *  callbackURL
+ *  participants { us them }
+ */
+
+export class Interaction {
+  private interactionFlow = {
+    [InteractionType.CredentialOfferRequest]: CredentialOfferFlow,
+    [InteractionType.CredentialRequest]: CredentialRequestFlow,
+    [InteractionType.Authentication]: AuthenticationFlow,
+  }
+  private interactionMessages: JSONWebToken<JWTEncodable>[] = []
+
+  public id: string
+  public ctx: BackendMiddleware
+  public flow!: Flow
+
+  // This is the channel through which the request (first token) came in.
+  public channel: InteractionChannel
+  public issuerSummary!: IdentitySummary
+
+  public constructor(
+    ctx: BackendMiddleware,
+    channel: InteractionChannel,
+    id: string,
+  ) {
+    this.ctx = ctx
+    this.channel = channel
+    this.id = id
+  }
+
+  public getMessages() {
+    return this.interactionMessages
+  }
+
+  // TODO Try to write a respond function that collapses these
+  public async createAuthenticationResponse() {
+    // TODO Abstract to getMessages * findByType
```

Comment:
Should probably be added as a small util function. This pattern repeats a number of times.

Reply:
:+1: